### PR TITLE
Check index or plugin name safety in cmd

### LINKED ISSUE #581: Some Krew commands susceptible to path manipulation
This is low-risk security issue so I'm opening it here.

While using `KREW_ROOT=~/.krew2`, I can still manipulate stuff in `~/.krew/` through the CLI by hijacking `PLUGIN` arguments with relative paths:

```
$ krew remove ../../.krew/receipts/whoami
Uninstalled plugin ../../.krew/receipts/whoami
```

This works even if `whoami` plugin is not installed in `~/.krew2` because we point to a valid receipt, which then gets deleted.

This happens because:
- `remove` command is not actually using `validation.IsSafePluginName` (like `LoadPluginByName` does) because it's operating on `Receipt`s
  - We actually added this check so that people don't use `name:` field in the plugin manifest to get random fs access.
- `environment.Paths` is vulnerable to manipulation because it doesn't do checks or return errors. I think this is by design

`krew index` cmds are not susceptible 
The methods who use LoadPluginByName fail correctly:

```
$ out/bin/krew-darwin_amd64 info ../../.krew/receipts/warp
F0401 11:02:03.677641   50768 root.go:79] failed to load plugin manifest: plugin name
"../.krew/receipts/warp" not allowed
```
```
out/bin/krew-darwin_amd64 upgrade ../../.krew/receipts/warp
Updated the local copy of plugin index.
F0401 11:04:52.552720   50879 root.go:79] failed to load the plugin manifest for plugin ../../.krew/receipts/warp: plugin name
"../.krew/receipts/warp" not allowed
```
```
out/bin/krew-darwin_amd64 install ../../.krew/receipts/warp
Updated the local copy of plugin index.
F0401 11:02:27.087157   50788 root.go:79] failed to load plugin "../../.krew/receipts/warp" from the index: plugin name
"../.krew/receipts/warp" not allowed
```

`krew index [add|remove]` are not susceptible to this as they use `IsValidIndexName` (although they do this check in pkg/indexoperations; not cmd/)

I propose we scan all cmds and validate plugin/index names with respective methods. The part that I'm not sure: should we do this in `cmd/` or `pkg/*`? Technically, cmd/ is a better place as it reduces test burden in lower layers.

### PR DESCRIPTION
**Pull request by ahmetb**
tl;dr: Some krew cmds are currently susceptible to path manipulation through positional arguments. Low-level machinery isn't responsible for this, so move them to cmd + add tests.

Moving plugin name or index name safety checks to cmd/ where they are given by
the user to the program. This way we shift the need for validation from being
unit tests of low-level machinery to the integration_test which tests
user-facing concerns.

> Note: Checks will fail until #582 is merged and this patch is rebased on it.

Fixes #581
/assign @corneliusweig
/assign @chriskim06

### CODE REVIEW

File: cmd/krew/cmd/install.go [hunk 1]
```diff
@@ -97,6 +98,10 @@ Remarks:
 
 			var install []pluginEntry
 			for _, name := range pluginNames {
+				if !validation.IsSafePluginName(name) {
+					return unsafePluginNameErr(name)
+				}
```
Comment (corneliusweig) [inline cmd/krew/cmd/install.go:103]: I'd feel better if we validate plugin names in batch upfront. One way we could achieve this is by changing the API to
```go
func IsSafePluginName(names ...string) bool { .. }
```

(same in several other places)

Also, it seems that `unsafePluginNameErr()` could be returned from this function (making it obsolete), so that we have instead:
```go
func CheckSafePluginName(names ...string) error { .. }
```

File: integration_test/uninstall_test.go [hunk 2]
```diff
@@ -75,3 +76,30 @@ func TestKrewRemove_ManifestRemovedFromIndex(t *testing.T) {
...
+	test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	test.Krew("install", validPlugin).RunOrFailOutput()
+
+	cases := []string{
+		`../index/` + validPlugin,
+		`..\index\` + validPlugin,
+		`../default/` + validPlugin,
+		`..\default\` + validPlugin,
+		`../receipts/` + validPlugin,
+		`..\receipts\` + validPlugin,
```
Comment (corneliusweig) [inline integration_test/uninstall_test.go:93]: How is `/receipt/` different from `/default/`? Why is this needed?
Otherwise, let's remove it.

File: integration_test/upgrade_test.go [hunk 3]
```diff
@@ -50,6 +50,31 @@ func TestKrewUpgrade(t *testing.T) {
...
+	defer cleanup()
+	test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+
+	cases := []string{
+		`../index/` + validPlugin,
+		`..\index\` + validPlugin,
+		`../default/` + validPlugin,
+		`..\default\` + validPlugin,
+		`index-name/sub-directory/plugin-name`,
+	}
```
Comment (corneliusweig) [inline integration_test/upgrade_test.go:65]: The list of test cases is repeated 3 times (if my question about `receipt` applies). Should we extract this to a central place?

File: integration_test/index_test.go [hunk 4]
```diff
@@ -45,6 +46,25 @@ func TestKrewIndexAdd(t *testing.T) {
...
+
+	cases := []string{"a/b", `a\b`, "../a", `..\a`}
+	expected := "invalid index name"
+
+	for _, c := range cases {
+		b, err := test.Krew("index", "add", c, constants.IndexURI).Run()
+		if err == nil {
+			t.Fatalf("%q: expected error", c)
+		} else if !strings.Contains(string(b), expected) {
+			t.Fatalf("%q: output doesn't contain %q: %q", c, expected, string(b))
```
Comment (chriskim06) [inline integration_test/index_test.go:63]: Small nit: `"%q: output doesn't contain %q: %q"` -> `"%q: output doesn't contain '%q': %q"`. This way it doesn't read as `a/b: output doesn't contain invalid index name...` but as `a/b: output doesn't contain 'invalid index name'...`

**Comment by corneliusweig** [comment]
Test failure seems legit:
```
You used fmt.Errorf; use pkg/errors.Errorf instead to preserve stack traces:
cmd/krew/cmd/uninstall.go:58:func unsafePluginNameErr(n string) error { return fmt.Errorf("plugin name %q not allowed", n) }
```

File: integration_test/index_test.go [hunk 5]
```diff
@@ -109,6 +129,24 @@ func TestKrewIndexRemove_ok(t *testing.T) {
...
+func TestKrewIndexRemove_unsafe(t *testing.T) {
+	skipShort(t)
+	test, cleanup := NewTest(t)
+	test = test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	defer cleanup()
+
+	expected := "invalid index name"
+	cases := []string{"a/b", `a\b`, "../a", `..\a`}
+	for _, c := range cases {
+		b, err := test.Krew("index", "add", c, constants.IndexURI).Run()
```
Comment (chriskim06) [inline integration_test/index_test.go:141]: I think this should be "remove"

File: integration_test/upgrade_test.go (hunk 3 above)
Comment (chriskim06) [reply integration_test/upgrade_test.go:65]: Also the test unsafe functions are all relatively similar (some of the error messaging is slightly different). The cases, expected error string, and arguments to `test.Krew` are the only differences. I think this could be useful as a `testutil` function like `test.AssertFailureWithUnsafeName` (not really sure what a good name would be)

File: integration_test/uninstall_test.go (hunk 2 above)
Comment (ahmetb) [reply integration_test/uninstall_test.go:93]: receipt/ is needed because we need to point this to a receipt file. and at the moment all these paths parseable as a Receipt

File: integration_test/upgrade_test.go (hunk 3 above)
Comment (ahmetb) [reply integration_test/upgrade_test.go:65]: I think duplication in tests aren't as bad, and each test actually has a slightly different things it's looking at, so I'm not entirely sure if the de-duplication helps.

File: integration_test/index_test.go (hunk 4 above)
Comment (ahmetb) [reply integration_test/index_test.go:63]: %q always adds `"` around the string, so this is not an issue.

File: integration_test/index_test.go (hunk 5 above)
Comment (ahmetb) [reply integration_test/index_test.go:141]: yep good catch.

File: cmd/krew/cmd/install.go (hunk 1 above)
Comment (ahmetb) [reply cmd/krew/cmd/install.go:103]: yeah although there's another method in `validation.go` that's geared towards plugin authors (and therefore it prints a regexp) as it runs during manifest validation tests.

    return errors.Errorf("the plugin name %q is not allowed, must match %q", name, safePluginRegexp.String())

that's why I kept it like this.

File: integration_test/index_test.go [hunk 6]
```diff
@@ -109,6 +129,24 @@ func TestKrewIndexRemove_ok(t *testing.T) {
...
+func TestKrewIndexRemove_unsafe(t *testing.T) {
+	skipShort(t)
+	test, cleanup := NewTest(t)
+	test = test.WithEnv(constants.EnableMultiIndexSwitch, 1).WithIndex()
+	defer cleanup()
+
+	expected := "invalid index name"
+	cases := []string{"a/b", `a\b`, "../a", `..\a`}
+	for _, c := range cases {
+		b, err := test.Krew("index", "remove", c, constants.IndexURI).Run()
```
Comment (chriskim06) [inline integration_test/index_test.go:141]: This last parameter isn't used in the index remove command

File: integration_test/index_test.go (hunk 4 above)
Comment (chriskim06) [reply integration_test/index_test.go:63]: TIL :smile:

**Comment by ahmetb** [comment]
Another round of reviews please. 🙏

File: integration_test/upgrade_test.go [hunk 7]
```diff
@@ -50,6 +50,31 @@ func TestKrewUpgrade(t *testing.T) {
...
+
+	expectedErr := `not allowed`
+	for _, c := range cases {
+		b, err := test.Krew("upgrade", c).Run()
```
Comment (chriskim06) [inline integration_test/upgrade_test.go:69]: Looks like you'd need https://github.com/kubernetes-sigs/krew/pull/582 before you can do this.

Comment (ahmetb) [reply integration_test/upgrade_test.go:69]: yep see the PR description :)

Comment (chriskim06) [reply integration_test/upgrade_test.go:69]: Whoops didn't see that haha! LGTM, I'll hold off for @corneliusweig to review.

File: integration_test/upgrade_test.go (hunk 3 above)
Comment (corneliusweig) [reply integration_test/upgrade_test.go:65]: It would be nice to have the unsafe names in a variable, so that it is easily re-used without the need to copy. But let's merge as-is and iterate from there.

File: cmd/krew/cmd/install.go (hunk 1 above)
Comment (corneliusweig) [reply cmd/krew/cmd/install.go:103]: I understand that you want to keep the API in `validation` consistent.

Maybe this helper can be put in `cmd/internal` instead? Just an idea. It's fine as it stands, but validating upfront would be even better IMO.

**Comment by corneliusweig** [comment]
@ahmetb Can you take a final look and rebase? Then I'll lgtm.

**Comment by ahmetb** [comment]
Rebased + test run https://github.com/ahmetb/krew/runs/559418218 actually has captured a legitimate bug. Posted a fix.

**Comment by corneliusweig** [comment]
/lgtm
/approve
Thanks for finding and fixing this loophole!
